Mathlib/CategoryTheory/Yoneda.lean

```diff
@@ -543,4 +544,201 @@ lemma isIso_of_yoneda_map_bijective {X Y : C} (f : X ⟶ Y)
   obtain ⟨g, hg : g ≫ f = 𝟙 Y⟩ := (hf Y).2 (𝟙 Y)
   exact ⟨g, (hf _).1 (by aesop_cat), hg⟩
 
+end YonedaLemma
+
+section CoyonedaLemma
+
+variable {C}
+
+/-- We have a type-level equivalence between natural transformations from the coyoneda embedding
+and elements of `F.obj X.unop`, without any universe switching.
+-/
+def coyonedaEquiv {X : C} {F : C ⥤ Type v₁} : (coyoneda.obj (op X) ⟶ F) ≃ F.obj X where
+  toFun η := η.app X (𝟙 X)
+  invFun ξ := { app := fun Y x ↦ F.map x ξ }
+  left_inv := fun η ↦ by
+    ext Y (x : X ⟶ Y)
+    dsimp
+    rw [← FunctorToTypes.naturality]
+    simp
+  right_inv := by intro ξ; simp
+
+theorem coyonedaEquiv_apply {X : C} {F : C ⥤ Type v₁} (f : coyoneda.obj (op X) ⟶ F) :
+    coyonedaEquiv f = f.app X (𝟙 X) :=
+  rfl
+
+@[simp]
+theorem coyonedaEquiv_symm_app_apply {X : C} {F : C ⥤ Type v₁} (x : F.obj X) (Y : C)
+    (f : X ⟶ Y) : (coyonedaEquiv.symm x).app Y f = F.map f x :=
+  rfl
+
+theorem coyonedaEquiv_naturality {X Y : Cᵒᵖ} {F : C ⥤ Type v₁} (f : coyoneda.obj X ⟶ F)
+    (g : Y ⟶ X) : F.map g.unop (coyonedaEquiv f) = coyonedaEquiv (coyoneda.map g ≫ f) := by
+  change (f.app X.unop ≫ F.map g.unop) (𝟙 X.unop) = f.app Y.unop (g.unop ≫ 𝟙 Y.unop)
+  rw [← f.naturality]
+  dsimp
+  simp
+
+lemma coyonedaEquiv_naturality' {X Y : C} {F : C ⥤ Type v₁} (f : coyoneda.obj (op X) ⟶ F)
```

Comment:
I think that the right statement is your `coyonedaEquiv_naturality'` and that `coyoneda_naturality` can be removed. (The idea is that in `coyonedaEquiv_naturality`, there are no `op` or `unop` of morphisms in the LHS.)